File: decode/js/rosetta-image-decode/src/internal/png.ts
import { PNG } from "pngjs";

import { DecodeError } from "../errors.js";
import type { DecodedImage } from "../types.js";

// Byte offset of bit-depth in a PNG IHDR chunk:
// 8 (magic) + 4 (chunk length) + 4 (chunk type "IHDR") + 4 (width) + 4 (height) = 24
const PNG_IHDR_BITDEPTH_OFFSET = 24;
const PNG_IHDR_COLORTYPE_OFFSET = 25;

function readPngHeaderBitDepth(bytes: Uint8Array): { bitDepth: number; colorType: number } | null {
  if (bytes.length < PNG_IHDR_COLORTYPE_OFFSET + 1) return null;
  return {
    bitDepth: bytes[PNG_IHDR_BITDEPTH_OFFSET]!,
    colorType: bytes[PNG_IHDR_COLORTYPE_OFFSET]!,
  };
}

export function decodePng(bytes: Uint8Array): DecodedImage {
  const buf = Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const header = readPngHeaderBitDepth(bytes);
  const is16bit = header?.bitDepth === 16;

  let png: PNG;
  try {
    // Use skipRescale only for 16-bit images so we can apply PIL-compatible conversion.
    // For 1/2/4/8-bit images, pngjs's normal scaling is already correct.
    png = PNG.sync.read(buf, is16bit ? { skipRescale: true } : {});
  } catch (e: any) {
    throw new DecodeError("corruptInput", "png", `PNG.sync.read failed: ${e.message ?? e}`);
  }

  const width = png.width;
  const height = png.height;
  const ct = (png as any).colorType as number;

  // Determine if the source image has an alpha channel.
  //   color type 0 = grayscale (no alpha)
  //   color type 2 = RGB (no alpha)
  //   color type 3 = paletted (alpha iff tRNS chunk present — pngjs sets `alpha` boolean)
  //   color type 4 = grayscale+alpha
  //   color type 6 = RGB+alpha
  const sourceAlphaFlag = (png as any).alpha === true;
  const sourceHasAlpha = ct === 4 || ct === 6 || (ct === 3 && sourceAlphaFlag);

  if (is16bit) {
    // pngjs data is a Uint16Array with skipRescale=true: one uint16 per channel (RGBA layout).
    // PIL conversion rules (from PIL/PngImagePlugin.py rawmode handling):
    //   colorType 0 (16-bit grayscale, no alpha): min(value, 255)
    //   all other 16-bit types:                   value >> 8  (top byte)
    const src = png.data as unknown as Uint16Array;
    const useClip = ct === 0; // only 16-bit grayscale uses min(v, 255)

    if (sourceHasAlpha) {
      // Output is RGBA (4 channels)
      const out = new Uint8Array(width * height * 4);
      for (let i = 0; i < width * height * 4; i++) {
        const v = src[i]!;
        out[i] = useClip ? Math.min(v, 255) : (v >> 8);
      }
      return { width, height, data: out, channels: 4, format: "png" };
    } else {
      // Output is RGB (3 channels); pngjs expands grayscale to R=G=B in RGBA buffer
      const out = new Uint8Array(width * height * 3);
      let si = 0;
      let di = 0;
      for (let i = 0; i < width * height; i++) {
        const r = src[si]!;
        const g = src[si + 1]!;
        const b = src[si + 2]!;
        // si+3 is alpha (always 65535 for no-alpha images) — skip
        out[di] = useClip ? Math.min(r, 255) : (r >> 8);
        out[di + 1] = useClip ? Math.min(g, 255) : (g >> 8);
        out[di + 2] = useClip ? Math.min(b, 255) : (b >> 8);
        si += 4;
        di += 3;
      }
      return { width, height, data: out, channels: 3, format: "png" };
    }
  }

  // 8-bit (or 1/2/4-bit expanded to 8-bit by pngjs): data is a Buffer with uint8 RGBA values.
  if (sourceHasAlpha) {
    return {
      width,
      height,
      data: new Uint8Array(png.data),
      channels: 4,
      format: "png",
    };
  }

  // Convert RGBA → RGB by dropping the alpha byte.
  // For grayscale source, pngjs has already expanded R=G=B; we just drop alpha.
  const rgbData = new Uint8Array(width * height * 3);
  let si = 0;
  let di = 0;
  for (let i = 0; i < width * height; i++) {
    rgbData[di] = png.data[si]!;
    rgbData[di + 1] = png.data[si + 1]!;
    rgbData[di + 2] = png.data[si + 2]!;
    si += 4;
    di += 3;
  }
  return {
    width,
    height,
    data: rgbData,
    channels: 3,
    format: "png",
  };
}

File: decode/rust/rosetta-image-decode/src/png.rs
use crate::error::{DecodeError, DecodeErrorKind};
use crate::types::{Channels, DecodedImage, Format};

use image::ColorType;
use std::io::Cursor;

pub(crate) fn decode_png(bytes: &[u8]) -> Result<DecodedImage, DecodeError> {
    // Peek at the magic bytes: if they don't match PNG, return UnsupportedFormat.
    // (detect_format already filters this before we're called, but guard defensively.)
    if bytes.len() < 8
        || bytes[0] != 0x89
        || bytes[1] != 0x50
        || bytes[2] != 0x4E
        || bytes[3] != 0x47
        || bytes[4] != 0x0D
        || bytes[5] != 0x0A
        || bytes[6] != 0x1A
        || bytes[7] != 0x0A
    {
        return Err(DecodeError::new(
            DecodeErrorKind::UnsupportedFormat,
            None,
            "not a PNG file",
        ));
    }

    let reader = image::ImageReader::with_format(
        Cursor::new(bytes),
        image::ImageFormat::Png,
    );

    let img = match reader.decode() {
        Ok(i) => i,
        Err(e) => {
            return Err(DecodeError::new(
                DecodeErrorKind::CorruptInput,
                Some(Format::Png),
                format!("image::decode failed: {}", e),
            ))
        }
    };

    let width = img.width() as usize;
    let height = img.height() as usize;
    let color = img.color();

    // Determine output channel count based on source color type.
    // PIL rules (matched by Java and Go ports):
    //   - Grayscale without alpha → RGB (replicate gray)
    //   - Grayscale with alpha (LA) → RGBA
    //   - RGB → RGB
    //   - RGBA → RGBA
    //   - Paletted without tRNS → RGB
    //   - Paletted with tRNS → RGBA
    // The image crate expands paletted images during decode; the resulting
    // ColorType reflects the expansion. We use has_alpha to decide.
    let has_alpha = matches!(
        color,
        ColorType::La8 | ColorType::La16 | ColorType::Rgba8 | ColorType::Rgba16
    );

    let channels = if has_alpha {
        Channels::Rgba
    } else {
        Channels::Rgb
    };

    // Special case: 16-bit single-channel grayscale (L16).
    // PIL opens these as mode 'I' (32-bit int) and clips to [0,255] on convert('RGB').
    // The image crate's to_rgb8() uses >> 8, which gives wrong results.
    // We must replicate: gray8 = min(raw_u16, 255), then expand to R=G=B.
    if matches!(color, ColorType::L16) {
        let gray16 = img.into_luma16();
        let raw = gray16.as_raw(); // &[u16] (native endian after decode)
        let pixel_count = width * height;
        let mut data = Vec::with_capacity(pixel_count * 3);
        for &v in raw.iter() {
            let clipped = v.min(255) as u8;
            data.push(clipped); // R
            data.push(clipped); // G
            data.push(clipped); // B
        }
        return Ok(DecodedImage {
            width,
            height,
            data,
            channels: Channels::Rgb,
            format: Format::Png,
        });
    }

    // Special case: 16-bit grayscale+alpha (La16).
    // PIL uses >> 8 for both gray and alpha channels.
    // image crate's to_rgba8() should handle this correctly (>> 8), but let's
    // be explicit via luma_alpha16 to guarantee byte-exactness.
    if matches!(color, ColorType::La16) {
        let la16 = img.into_luma_alpha16();
        let raw = la16.as_raw(); // [gray, alpha, gray, alpha, ...]
        let pixel_count = width * height;
        let mut data = Vec::with_capacity(pixel_count * 4);
        for chunk in raw.chunks_exact(2) {
            let gray = (chunk[0] >> 8) as u8;
            let alpha = (chunk[1] >> 8) as u8;
            data.push(gray);  // R
            data.push(gray);  // G
            data.push(gray);  // B
            data.push(alpha); // A
        }
        return Ok(DecodedImage {
            width,
            height,
            data,
            channels: Channels::Rgba,
            format: Format::Png,
        });
    }

    // Special case: 16-bit RGB (Rgb16).
    // PIL uses >> 8. image crate's to_rgb8() uses >> 8 as well, but let's verify
    // by using the raw buffer directly to be safe.
    if matches!(color, ColorType::Rgb16) {
        let rgb16 = img.into_rgb16();
        let raw = rgb16.as_raw(); // [r, g, b, r, g, b, ...]
        let pixel_count = width * height;
        let mut data = Vec::with_capacity(pixel_count * 3);
        for &v in raw.iter() {
            data.push((v >> 8) as u8);
        }
        return Ok(DecodedImage {
            width,
            height,
            data,
            channels: Channels::Rgb,
            format: Format::Png,
        });
    }

    // Special case: 16-bit RGBA (Rgba16).
    // PIL uses >> 8 for all channels.
    if matches!(color, ColorType::Rgba16) {
        let rgba16 = img.into_rgba16();
        let raw = rgba16.as_raw(); // [r, g, b, a, ...]
        let pixel_count = width * height;
        let mut data = Vec::with_capacity(pixel_count * 4);
        for &v in raw.iter() {
            data.push((v >> 8) as u8);
        }
        return Ok(DecodedImage {
            width,
            height,
            data,
            channels: Channels::Rgba,
            format: Format::Png,
        });
    }

    // For all remaining 8-bit types (L8, Rgb8, Rgba8, and paletted which the
    // image crate expands to Rgb8 or Rgba8), use the standard conversions.
    // image crate's to_rgb8()/to_rgba8() match PIL for these types.
    let data: Vec<u8> = if has_alpha {
        img.to_rgba8().into_raw()
    } else {
        img.to_rgb8().into_raw()
    };

    Ok(DecodedImage {
        width,
        height,
        data,
        channels,
        format: Format::Png,
    })
}
